src/devices/machine/ie15_kbd.cpp

```diff
@@ -29,83 +32,39 @@ ie15_keyboard_device::ie15_keyboard_device(const machine_config &mconfig, const
 {
 }
 
-uint8_t ie15_keyboard_device::row_number(uint32_t code)
+INPUT_CHANGED_MEMBER(ie15_keyboard_device::dip_changed)
 {
-	int row = 0;
-
-	while (code >>= 1) { ++row; }
-
-	return row;
+	m_sdv_cb(newval);
 }
 
-uint16_t ie15_keyboard_device::keyboard_handler(uint16_t last_code, uint8_t *scan_line)
+void ie15_keyboard_device::key_make(uint8_t row, uint8_t column)
 {
-	int i;
-	uint32_t code = 0;
 	uint16_t key_code = 0;
-	uint16_t retVal = 0;
-	uint8_t ctrl  = BIT(m_io_kbdc->read(), 0);
-	uint8_t shift = BIT(m_io_kbdc->read(), 1);
+	int ctrl  = BIT(m_io_kbdc->read(), 0);
+	int shift = BIT(m_io_kbdc->read(), 1);
 
-	i = *scan_line;
-	{
-		if (i<4)
-			code = m_io_kbd[i]->read();
-
-		if (code != 0)
-		{
-			if (!m_ruslat)
-				i += 4;
-			if (shift)
-				i += 8;
-			i <<= 4;
-			if (code < 0x10000) {
-				key_code = m_rom [i + row_number(code)];
-			} else {
-				key_code = m_rom [i + row_number(code >> 16)*2 + 256];
-			}
-			if (ctrl) {
-				key_code &= 0x1f;
-			}
-			retVal = key_code;
-		} else {
-			*scan_line += 1;
-			if (*scan_line==4) {
-				*scan_line = 0;
-			}
-		}
-	}
+	if (m_ruslat)
+		row += 4;
+	if (shift)
+		row += 8;
 
-	// setup mode toggle
-	if (BIT(m_io_kbdc->read(), 2)) {
-		retVal |= (IE_KB_SDV << 8);
-	}
-	// aux keyboard flag
-	if (code & 0xe0e000) {
-		retVal |= (IE_KB_DK << 8);
-	}
-	// shift
-	if (shift) {
-		retVal |= (IE_KB_NR << 8);
-	}
-	// SI/SO handling
-	if ((key_code & 127) == IE_KB_SI) {
-		m_ruslat = 0;
-	}
-	if ((key_code & 127) == IE_KB_SO) {
-		m_ruslat = 1;
+	if (column < 16)
+		key_code = m_rom [(row << 4) + column];
+	else
+		key_code = m_rom [(row << 4) + (column - 16) * 2 + 256];
+
+	if (ctrl)
+		key_code &= 0x1f;
+
+	// XXX aux keyboard flag
+	if ((column > 12 && column < 16) || (column > 20 && column < 24))
+	{
+		key_code |= (IE_KB_DK << 8);
 	}
 
-	return retVal;
-}
+	logerror("key_make row %d col %d key_code %04X (%02X)\n", row, column, key_code, key_code & 0x7f);
```

Comment:
Did you intend to make it always log keypresses, or should this be using a logmacro.h thing?

Reply:
I should delete that :)